File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/api/provider/action/NativeActionProviderRegistry.kt
package io.nativeblocks.runtime.api.provider.action

internal object NativeActionProviderRegistry {

    private val providerMap = mutableMapOf<String, NativeActionProvider>()
    private val lock = Any()

    fun getOrCreate(instanceName: String): NativeActionProvider {
        synchronized(lock) {
            return providerMap.getOrPut(instanceName) {
                NativeActionProvider()
            }
        }
    }

    fun remove(instanceName: String) {
        synchronized(lock) {
            providerMap.remove(instanceName)
        }
    }

}

File: platforms/ios/runtime/Sources/NativeblocksRuntime/api/provider/action/NativeActionProviderRegistry.swift
import Foundation

internal class NativeActionProviderRegistry {
    private static var providerMap: [String: NativeActionProvider] = [:]
    private static let lock = NSLock()

    static func getOrCreate(_ instanceName: String) -> NativeActionProvider {
        lock.lock()
        defer { lock.unlock() }

        if let existingProvider = providerMap[instanceName] {
            return existingProvider
        } else {
            let newProvider = NativeActionProvider()
            providerMap[instanceName] = newProvider
            return newProvider
        }
    }

    static func remove(_ instanceName: String) {
        lock.lock()
        defer { lock.unlock() }
        providerMap.removeValue(forKey: instanceName)
    }

}
